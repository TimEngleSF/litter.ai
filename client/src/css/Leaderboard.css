.lb-container {
  height: 90vh;
  width: 42rem;
  max-width: 100vw;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  margin-top: 10%;
}

.lb-container
.lb-user-stats {
  width: 80%;
  height: 5vh;
  display: flex;
  align-items: center;
  justify-content: space-around;
  border: 0.2rem var(--main-green) solid;
  border-radius: 5rem;
  margin-bottom: 2rem;
}

.lb-container
.lb-table {
  width: 95%;
  height: 60vh;
  border: 0.15rem var(--main-green) solid;
  border-spacing: inherit;
  text-align: center;
}

.lb-container
.lb-table tr:nth-of-type(even) {
  background: var(--lb-gray);
}

.lb-container
.lb-table th + th {
  border-left: 0.1rem var(--white)solid;
}

.lb-container
.lb-table td + td {
  border-left: 0.1rem var(--main-green)solid;
}

.lb-container
.lb-table
.lb-header {
  background: var(--main-green);
  color: var(--white);
}

.lb-container
.lb-table
.lb-header th:nth-child(1) {
  width: 15%;
}

.lb-container
.lb-table
.lb-header th:nth-child(2) {
  height: 2.5rem;
  width: 20rem;
  text-align: left;
  padding-left: 1rem;
}

.lb-container
.lb-table
.lb-header th:nth-child(3) {
  width: 25%;
}

.lb-container
.lb-name {
  text-align: left;
  padding-left: 1rem;
}

.lb-container
.profile-buttons {
  max-width: 100%;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: space-between;
  margin-bottom: 1rem;
  margin-top: 10%;
}

.lb-container
.profile-buttons
.button-home {
  background-color: var(--white);
}

.lb-container
.profile-buttons button {
  width: 30vw;
  max-width: 653px;
  height: 6vh;
  margin: 0.3rem 0;
  border: none;
  border-radius: 10rem;
  font-size: 1.2rem;
}

.lb-container
.profile-buttons button:hover {
  cursor: pointer;
  background-color: var(--main-green);
  color: var(--white);
	box-shadow: 0.25em 0.25em .75em -.5em var(--shadow-color);
  font-weight: bold;
}

@media only screen and (max-width: 600px)  {
  .lb-container {
    grid-area: 5 / 1 / span 1 / span 1;
  }

  .lb-container
  .lb-table
  .lb-header {
    font-size: 0.7rem;
  }

  .lb-container
  .lb-table
  .lb-header th:nth-child(3) {
    width: 30%;
  }

  .lb-container
  .lb-table tbody {
    font-size: 0.8rem;
  }

  .lb-container
  .profile-buttons button {
  width: 85vw;
  }
}

@media only screen and (min-width: 601px) and (max-width: 992px) {
  .lb-container {
    grid-area: 5 / 2 / span 1 / span 1;
  }

  .lb-container
  .profile-buttons button {
  width: 85vw;
  }
}

@media only screen and (min-width: 993px) {
  .lb-container {
    grid-area: 1 / 2 / span 10 / span 1;
  }
}