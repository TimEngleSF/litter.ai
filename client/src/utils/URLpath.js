export const aiEndpoint = process.env.REACT_APP_AI_KEY || 'https://aii.littersort.ai';

export default function URLpath(endpoint, extension) {
  const deployed = process.env.REACT_APP_API_KEY || 'https://api.littersort.ai';
  // const local = 'http://localhost:3001';
  const link = deployed;
  return extension
    ? `${link}/${endpoint}/${extension}`
    : `${link}/${endpoint}`;
}
