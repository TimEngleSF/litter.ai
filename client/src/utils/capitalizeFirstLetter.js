export const capitalizeList = (list) => (
  Object.keys(list).map(
    (key) => [key.charAt(0).toUpperCase() + key.slice(1), list[key]],
  )
);

export const capitalizeWord = (word) => (
  word.charAt(0).toUpperCase() + word.slice(1)
);
