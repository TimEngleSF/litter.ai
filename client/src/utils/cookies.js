import Cookies from 'js-cookie';

export default function loginAuth(loginToken) {
  const token = Cookies.get('authToken');
  if (token) return token;
  Cookies.set('authToken', loginToken, { expires: 7 });
  return token;
}
