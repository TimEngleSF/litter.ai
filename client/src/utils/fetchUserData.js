/* eslint-disable consistent-return */
import Cookies from 'js-cookie';
import fetchData from './fetch';

export async function fetchLogin(body) {
  try {
    const response = await fetchData('login', 'POST', body);
    if (response.token) {
      Cookies.set('authToken', response.token, { expires: 7 });
      Cookies.set('username', response.user.displayUsername, { expires: 7 });
    }
    return response;
  } catch (err) {
    console.error(err);
  }
}

export async function fetchLogOut() {
  try {
    await fetchData('logout', 'POST');
    await Cookies.remove('authToken');
    await Cookies.remove('username');
  } catch (err) {
    console.error(err);
  }
}

export async function fetchLeaderboardData(path) {
  try {
    const response = await fetchData(path, 'GET');
    return response;
  } catch (err) {
    console.error(err);
  }
}

export async function fetchProfileData(user) {
  try {
    const response = await fetchData(`photo/${user}`, 'GET');
    return response;
  } catch (err) {
    console.error(err);
  }
}

export async function fetchRegister(body) {
  try {
    const response = await fetchData('register', 'POST', body);
    if (response.status === 'pending') {
      const loginBody = {
        email: body.email,
        password: body.password,
      };
      const loginResponse = await fetchLogin(loginBody);
      return loginResponse;
    }
    return response;
  } catch (err) {
    console.error(err);
  }
}
