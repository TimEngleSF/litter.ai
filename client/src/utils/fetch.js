/* eslint-disable consistent-return */
import Cookies from 'js-cookie';
import URLpath from './URLpath';

export default async function fetchData(path, method, body) {
  const token = Cookies.get('authToken');
  const properties = {
    method,
    headers: {
      'Content-Type': 'application/json',
    },
  };
  // Leaderboard queries for the user's stats if they are logged in
  if (token) { properties.headers.Authorization = `Bearer ${token}`; }
  // Body is the email and password for logging in
  if (body) { properties.body = JSON.stringify(body); }

  try {
    const res = await fetch(URLpath(path), properties);
    const response = await res.json();
    return response;
  } catch (err) {
    console.error(err);
  }
}
