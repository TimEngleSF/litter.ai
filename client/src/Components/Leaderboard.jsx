/* eslint-disable react/jsx-no-useless-fragment */
import React, { useState, useEffect } from 'react';
import { Link } from 'react-router-dom';
import Dropdown from './Dropdown';
import { fetchLeaderboardData } from '../utils/fetchUserData';
import '../css/Leaderboard.css';

export default function Leaderboard() {
  const [leaderboardData, setLeaderboardData] = useState([]);
  const [dropdownSelection, setDropdownSelection] = useState('total');
  const [userRank, setUserRank] = useState(null);
  const [userItemCount, setUserItemCount] = useState(null);

  // Creates a data row for each user with their rank, username, and total uploads
  const renderTable = (user, index) => (
    <tr key={index}>
      <td>{index + 1}</td>
      <td className="lb-name">{user.username}</td>
      <td>{user.itemCount}</td>
    </tr>
  );

  // Sorts users by total uploads
  useEffect(() => {
    // Data for the total uploads is a different link than the categories
    const path = dropdownSelection === 'total' ? 'leaderboard' : `leaderboard/${dropdownSelection}`;

    // Authenticates user and grabs user's leaderboard data
    const fetchData = async () => {
      const response = await fetchLeaderboardData(path);
      setLeaderboardData(response.leaderboard);
      setUserRank(response.userRank);
      setUserItemCount(response.userItemCount);
    };
    fetchData();
  }, [dropdownSelection]);

  return (
    <>
      { leaderboardData.length > 0 && (
        <div className="lb-container main-container">
          <h1>Leaderboard</h1>
          { userRank && (
          <div className="lb-user-stats">
            <h3>
              Your rank:&nbsp;
              {userRank > 0 ? userRank : 'No photos'}
            </h3>
            <h3>
              {dropdownSelection}
              :&nbsp;
              {userItemCount}
            </h3>
          </div>
          )}
          <table className="lb-table">
            <thead>
              <tr className="lb-header">
                <th scope="col">Rank</th>
                <th scope="col">Name</th>
                <th scope="col">
                  <Dropdown
                    setLBCategory={setDropdownSelection}
                    defaultValue={null}
                    aria-label="Dropdown"
                  />
                </th>
              </tr>
            </thead>
            <tbody>
              {leaderboardData.map((user, index) => (renderTable(user, index)))}
            </tbody>
          </table>
          <div className="lb-buttons lower-buttons">
            <Link to="/capture">
              <button type="button">
                Capture Picture
              </button>
            </Link>
            <Link to="/">
              <button className="button-home" type="button">
                Home
              </button>
            </Link>
          </div>
        </div>
      )}
    </>
  );
}
