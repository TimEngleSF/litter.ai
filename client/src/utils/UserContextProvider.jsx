import React, { useState, useMemo } from 'react';
import { fetchLogin, fetchLogOut } from './fetchUserData';
import userContext from './UserContext';

export default function UserContextProvider({ children }) {
  const [userData, setUserData] = useState({});

  const login = async (loginData) => {
    try {
      const authUserData = await fetchLogin(loginData);
      setUserData(authUserData);
    } catch (err) {
      console.error(err);
    }
  };

  const contextValue = useMemo(() => ({
    userData,
    setUserData,
    login,
  }), [userData]);

  return (
    <userContext.Provider value={contextValue}>
      {children}
    </userContext.Provider>
  );
}
