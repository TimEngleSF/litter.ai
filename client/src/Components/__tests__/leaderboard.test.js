/* eslint-disable testing-library/no-unnecessary-act */
/* eslint-disable react/jsx-filename-extension */
import React from 'react';
import renderer from 'react-test-renderer';
import {
  render,
  act,
  screen,
  fireEvent,
  waitFor,
} from '@testing-library/react';
import { MemoryRouter as Router } from 'react-router-dom';
import Leaderboard from '../Leaderboard';
import * as fetchUserData from '../../utils/fetchUserData';
import {
  mockTotalUploads,
  mockPlasticUploads,
  mockMetalUploads,
} from '../../MockData/mockLeaderboardData';

describe('Leaderboard component', () => {
  const leaderboard = <Router><Leaderboard /></Router>;
  // Function to simulate an API call
  const mockCall = async (data) => {
    jest.spyOn(fetchUserData, 'fetchLeaderboardData').mockResolvedValue(data);
  };

  // Start each test with total top 10 API call; reset mock after each test
  beforeEach(() => { mockCall(mockTotalUploads); });
  afterEach(() => { jest.restoreAllMocks(); });

  // Creates a snapshot
  test('Leaderboard matches the current snapshot', async () => {
    const tree = await act(async () => renderer.create(leaderboard));
    // Update tree because it renders nulls on initial load
    await act(() => tree.update(leaderboard));
    expect(tree.toJSON()).toMatchSnapshot();
  });

  test('Changes category from total to compost when clicking on the dropdown', async () => {
    await act(() => render(leaderboard));
    // Clicks on dropdown
    fireEvent.mouseDown(screen.getByText('Total'));
    // Changes dropdown to 'Compost'
    await act(() => fireEvent.click(screen.getByText('Compost')));
    // Checking if 'Compost' is now the header on the dropdown
    expect(screen.getByText('Compost')).toBeInTheDocument();
  });

  test('Dropdown option text and background changes color when hovered', async () => {
    await act(() => render(leaderboard));
    // Clicks on dropdown
    fireEvent.mouseDown(screen.getByText('Total'));
    // Hover over new category
    fireEvent.mouseEnter(screen.getByText('Cardboard'));
    // Checking if 'Cardboard' has the hovered color and background color
    const mouseOverCardboard = getComputedStyle(screen.getByText('Cardboard'));
    expect(mouseOverCardboard.color).toBe('rgb(0, 0, 0)');
    expect(mouseOverCardboard.backgroundColor).toBe('rgb(116, 204, 103)');
  });

  test('New data is rendered when a dropdown category is changed', async () => {
    await act(() => render(leaderboard));
    // Change category and data to metal
    await act(() => { mockCall(mockMetalUploads); });

    fireEvent.mouseDown(screen.getByText('Total'));
    act(() => fireEvent.click(screen.getByText('Metal')));
    await waitFor(() => { expect(screen.getByText(/No Photos/i)).toBeInTheDocument(); });
    // Checks if a specific user is in the document
    await waitFor(() => { expect(screen.getByText('lucious_senger10')).toBeInTheDocument(); });
    // Change category and data to plastic
    fireEvent.mouseDown(screen.getByText('Metal'));
    act(() => { mockCall(mockPlasticUploads); });
    act(() => fireEvent.click(screen.getByText('Plastic')));
    // Checks if top metal user is no longer there and if a top plastic user is now visible
    await waitFor(() => { expect(screen.queryByText('lucious_senger10')).not.toBeInTheDocument(); });
    await waitFor(() => { expect(screen.getByText('alejandra31')).toBeInTheDocument(); });
  });
});
