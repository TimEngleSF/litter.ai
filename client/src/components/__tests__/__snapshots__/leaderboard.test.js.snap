// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Leaderboard component Leaderboard matches the current snapshot 1`] = `
<div
  className="lb-container main-container"
>
  <h1>
    Leaderboard
  </h1>
  <div
    className="lb-user-stats"
  >
    <h3>
      Your rank: 
      3
    </h3>
    <h3>
      total
      : 
      23423
    </h3>
  </div>
  <table
    className="lb-table lb-table-user"
    data-testid="lb-table"
  >
    <thead>
      <tr
        className="lb-header"
      >
        <th
          scope="col"
        >
          Rank
        </th>
        <th
          scope="col"
        >
          Name
        </th>
        <th
          scope="col"
        >
          <div>
            <div
              className=" css-b62m3t-container"
              onKeyDown={[Function]}
            >
              <span
                className="css-1f43avz-a11yText-A11yText"
                id="react-select-2-live-region"
              />
              <span
                aria-atomic="false"
                aria-live="polite"
                aria-relevant="additions text"
                className="css-1f43avz-a11yText-A11yText"
                role="log"
              />
              <div
                className=" css-jc6pza-control"
                onMouseDown={[Function]}
                onTouchEnd={[Function]}
              >
                <div
                  className=" css-1fdsijx-ValueContainer"
                >
                  <div
                    className=" css-ayhps2-singleValue"
                  >
                    Total
                  </div>
                  <input
                    aria-activedescendant=""
                    aria-autocomplete="list"
                    aria-expanded={false}
                    aria-haspopup={true}
                    aria-readonly={true}
                    className="css-mohuvp-dummyInput-DummyInput"
                    disabled={false}
                    id="react-select-2-input"
                    inputMode="none"
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    role="combobox"
                    tabIndex={0}
                    value=""
                  />
                </div>
                <div
                  className=" css-1hb7zxy-IndicatorsContainer"
                >
                  <span
                    className=" css-1u9des2-indicatorSeparator"
                  />
                  <div
                    aria-hidden="true"
                    className=" css-1s7cnte-indicatorContainer"
                    onMouseDown={[Function]}
                    onTouchEnd={[Function]}
                  >
                    <svg
                      aria-hidden="true"
                      className="css-tj5bde-Svg"
                      focusable="false"
                      height={20}
                      viewBox="0 0 20 20"
                      width={20}
                    >
                      <path
                        d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                      />
                    </svg>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          1
        </td>
        <td
          className="lb-name"
        >
          briana30
        </td>
        <td />
      </tr>
      <tr>
        <td>
          2
        </td>
        <td
          className="lb-name"
        >
          forrest_windler53
        </td>
        <td />
      </tr>
      <tr>
        <td>
          3
        </td>
        <td
          className="lb-name"
        >
          alvera.kub
        </td>
        <td />
      </tr>
      <tr>
        <td>
          4
        </td>
        <td
          className="lb-name"
        >
          alanna98
        </td>
        <td />
      </tr>
      <tr>
        <td>
          5
        </td>
        <td
          className="lb-name"
        >
          amparo.paucek
        </td>
        <td />
      </tr>
      <tr>
        <td>
          6
        </td>
        <td
          className="lb-name"
        >
          granville.nicolas
        </td>
        <td />
      </tr>
      <tr>
        <td>
          7
        </td>
        <td
          className="lb-name"
        >
          sigrid.wilderman9
        </td>
        <td />
      </tr>
      <tr>
        <td>
          8
        </td>
        <td
          className="lb-name"
        >
          shemar81
        </td>
        <td />
      </tr>
      <tr>
        <td>
          9
        </td>
        <td
          className="lb-name"
        >
          rocio.hudson
        </td>
        <td />
      </tr>
      <tr>
        <td>
          10
        </td>
        <td
          className="lb-name"
        >
          favian88
        </td>
        <td />
      </tr>
    </tbody>
  </table>
</div>
`;
